### PR TITLE
Switch IceInternal::Buffer underlying type to std::byte

### PR DESCRIPTION
**Pull request by externl**
This PR implements #1825, namely

- Update IceInternal::Buffer to use std::byte
- Update Slice to map `sequence<byte>` to `std::vector<std::byte>` by default

### CODE REVIEW

File: cpp/src/Ice/InputStream.cpp [hunk 1]
```diff
@@ -1174,12 +1180,18 @@ Ice::InputStream::read(wstring& v)
             if (_instance)
             {
                 const WstringConverterPtr& wstringConverter = _instance->getWstringConverter();
-                wstringConverter->fromUTF8(i, i + sz, v);
+                wstringConverter->fromUTF8(
+                    reinterpret_cast<const uint8_t*>(i),
```
Comment (bernardnormier) [inline cpp/src/Ice/InputStream.cpp:1184]: We should fix fromUTF8 to take bytes like other APIs.

File: cpp/src/Ice/OutputStream.cpp [hunk 2]
```diff
@@ -39,7 +39,7 @@ namespace
                 //
                 // Return unused bytes
                 //
-                _stream.resize(static_cast<size_t>(firstUnused - _stream.b.begin()));
+                _stream.resize(static_cast<size_t>(firstUnused - reinterpret_cast<uint8_t*>(_stream.b.begin())));
```
Comment (bernardnormier) [inline cpp/src/Ice/OutputStream.cpp:42]: We should fix the UTF8Buffer API to use bytes.

File: cpp/src/slice2cpp/CPlusPlusUtil.cpp [hunk 3]
```diff
@@ -49,12 +49,20 @@ namespace
             if (seqType == "%array")
             {
                 BuiltinPtr builtin = dynamic_pointer_cast<Builtin>(seq->type());
-                string s = typeToString(
-                    seq->type(),
-                    false,
-                    scope,
-                    seq->typeMetaData(),
-                    typeCtx | (inWstringModule(seq) ? TypeContext::UseWstring : TypeContext::None));
+                string s;
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply cpp/src/slice2cpp/CPlusPlusUtil.cpp:52]: I think that's fine.

**Review by bernardnormier — APPROVED** [review]
We should fix the UTF8-related code in a follow-up PR, and remove some related reinterpret_cast.

**Review by pepone — APPROVED** [review]

**Pull request merged**